keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -143,29 +144,51 @@ The CRD `v1` API will be the same as the `v1beta1` but with all changes to the A
 
 ### Integration tests for GA
 
-TODO: complete this list
-
-* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
-  * TODO
-* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
-  * TODO
+* Ensure CRD themselves are v1<->v1beta1 round trippable
 * CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
   * Ensure what is persisted in etcd matches the storage version
   * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
   * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
   * Ensure `spec.served` is respected
+* Tests for malformed inputs and errors (system should return sensible errors, not panic, and recover appropriately)
+  * TODO
 
 ### e2e tests for GA
 
-* Custom Resources should be readable and writable at all available versions (test for round trip-ability)
 * A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
 
 ### Scale Targets for GA
 
-* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+The scale target for GA of custom resources are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details):
+
+| SLI | SLO |
+| --- | --- |
+| Latency<sup>[1](#footnote1)</sup> of mutating<sup>[2](#footnote2)</sup> API calls for single objects for every (resource, verb) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, verb) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day <= 1s |
+| Latency<sup>[1](#footnote1)</sup> of non-streaming read-only<sup>[3](#footnote3)</sup> API calls for every (resource, scope<sup>[4](#footnote4)</sup>) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, scope) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day (a) <= 1s if `scope=resource` (b) <= 5s if `scope=namespace` (c) <= 30s if `scope=cluster` |
+
+And with these thresholds:
+
+| Quantity | Threshold scope=namespace | Threshold: scope=cluster |
+| --- | --- | --- |
+| Custom Resource objects per Definition | 1000 | 5000 |
+| Custom Resource Definitions | 100 | 500 |
```

Comment:
BTW - 500*5000 seems like quite a lot. Maybe we should also have total CRs/cluster limit.

Reply:
CRD with scope=namespace = CRD for CRs which are namespaced

Reply:
Is the scope=namespace value per namespace or all namespaces summed up? What is the list API call we are talking about? Do we talk about the non-namespaced LIST for namespaced resources?

Reply:
> CRD with scope=namespace = CRD for CRs which are namespaced

Ahh - I didn't know that. Thanks.

> Is the scope=namespace value per namespace or all namespaces summed up? 

Per-namespace.

There is one potential problem with that definition then, because for all other resources we know whether they are namespace-scope or not. Which basically translates to:
- scope=namespace => limit per a single namespace
- scope=cluster => total number of all objects across all namespaces

> Do we talk about the non-namespaced LIST for namespaced resources?

Yeah - that is how it was up until now. We may clarify it a bit more for CRDs given your first sentence.

Reply:
> 500*5000 seems like quite a lot

@wojtek-t I agree, especially when each CR object can be of 50kb size. May I ask are the [scalability thresholds](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds) defined as _"SLOs being satisfied when all dimensions are being pushed at the same time"_, or _"... when one dimension is pushed and the other dimensions are light-loaded"_? My interpretation was the latter, and we know that 
> As you move farther along one dimension, your cross-section wrt other dimensions gets smaller

Reply:
will it make sense to start with something basic, focusing on one resource (assuming we only have one CRD with lots of CRs)? e.g. 

| Quantity | Threshold scope=namespace | Threshold: scope=cluster |
| --- | --- | --- |
| Namespaced Custom Resource objects per Definition |  |  |
| Cluster-scoped Custom Resource objects per Definition | n/a |  |
| Custom Resource Definitions | n/a |  |

Reply:
Yeah - it's closer to the latter. I mean there are dimensions that you can all push at the same time, but not all of them.

Regarding the table above - yes that's exactly how I would see that.

Reply:
I'm thinking of explaining this with some text like: "For `scope: Namespaced` custom resource definitions, the scope=namespace threshold indicates how many custom resource objects may be in each namespace, and the scope=cluster threshold indicates how many custom resource objects may be in the cluster total. For `scope: Cluster` custom resource definitions, only the scope=cluster threshold applies."